File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.cpp
class Solution {
public:
    string smallestSubsequence(string s) {
                vector<int> last_occurence(26, -1);    
        for(int i=0; i<s.length(); i++) {
            last_occurence[s[i]-'a'] = i;
        }
        
        vector<bool> visited(26, false);
        stack<char> st;
        
        for(int i=0; i<s.length(); i++) {
            if(visited[s[i]-'a']) continue; // if element is already present
            while(!st.empty() && last_occurence[st.top()-'a'] > i /* this element is present ahead */ && st.top() > s[i]) {
                visited[st.top()-'a'] = false;
                st.pop();
            }  
            
            visited[s[i]-'a'] = true;
            st.push(s[i]);
        }
        
        string res = "";
        while(!st.empty()) {
            res.push_back(st.top());
            st.pop();
        }
        reverse(res.begin(), res.end());
        
        return res;
    
    }
};

File: 1081-smallest-subsequence-of-distinct-characters/1081-smallest-subsequence-of-distinct-characters.java
class Solution {
    public String smallestSubsequence(String s) {
     int[] arr=new int[26];
        for(int i=0;i<s.length();i++)
        {
            arr[s.charAt(i)-'a']=i;
        }
        boolean[] visited=new boolean[26];
        Deque<Integer> st=new ArrayDeque<>();
        for(int i=0;i<s.length();i++)
        {
            int j=s.charAt(i)-'a';
            if(visited[j])
                continue;
            while(!st.isEmpty() && st.peek()>j && arr[st.peek()]>i)
            {
                visited[st.pop()]=false;
            }
            st.push(j);
            visited[j]=true;
        }
        StringBuilder str=new StringBuilder();
        while(!st.isEmpty())
        {
            str.append((char)(st.pop()+'a'));
        }
        return str.reverse().toString();
    }
}